rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,410 @@
+with the inner workings of the system.
+
+## User-facing explanation
+Both of the current components' public interface will be made *read-only*.
+Structural mutations to the hierarchy: de-parenting, re-parenting, and moving
+children from one parent to another, etc. can only be done via commands. This
+defers all changes to the hierarchy until the next command buffer flush.
+This enforces a globally consistent view at any given time. Delaying any
+modifications to a global synchronization point, much like component removal.
+Both components will not publicly constructible and must be made via the
+dedicated hierarchy management commands.
+
+`PreviousParent` will also be removed and replaced with `ChildAdded`,
+`ChildRemoved` and `ChildMoved` events instead. These will signal changes in the
+hierachy instead of relying on change detection on `Parent` and `PreviousParent`.
+
+## Implementation strategy
+This design attempts to minimize the changes made to the current design while
+addressing the issues enumerated above. The core changes to be made are as
+follows:
+
+ - Make `Parent` and `Children` (mostly) publicly immutable.
+ - Make mutations to the hierarchy rely solely on commands.
+ - Update hierarchy commands to automatically update both `Parent` and
+   `Children` to remain consistent with each other.
+ - Remove the `parent_update_system`.
+ - Remove `PreviousParent`.
+ - Register several hierarchy events that fires whenever time a new child is
+   added or removed from a parent.
+ - Create a `HiearchyQuery` custom SystemParam for more easily iterating over
+   components in the hierarchy.
+
+This change will repurpose the following existing `EntityCommand` extension
+functions for this purpose:
+
+ - `add_child`
+ - `push_children`
+ - `insert_children`
+ - `remove_child` (this will be added as a wrapper around `remove_children`)
+ - `remove_children`
+
+An additional `set_parent(Option<Entity>)` command should also be added as a
+child-side way of altering the hierarchy, defaulting to adding the entity as the
+last child to the provided parent if provided, and detaching the child from it's
+parent if `None`.
+
+### HierarchyQuery
+An extension to this design is to create dedicated `HierarchyQuery<Q, F>`, which
+is a custom `SystemParam` that wraps multiple `Query` objects to simplify
+querying for items in the hierarchy:
+
+ - `roots(_mut)` - returns an iterator over entities that match the provided
+   query parameters that do not have a parent.
+ - `parent(_mut)` - returns single result from the parent of an entity.
+ - `iter_children(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
```

Comment:
Currently, they sound like they do the exact same thing.

```suggestion
 - `iter_children(_mut)` - returns an iterator over `(parent, child)` results
   from a query parameter. This does not return Children of Children.
 - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
   from a query parameter. This does a depth-first traversal over the hierarchy
   starting at some entity.
```

As for `iter_descendants` which `depth-first traversal` does it perform?
Inorder (Left, Root, Right), Preorder (Root, Left, Right) or Postorder (Left, Right, Root)?
I would expect Preorder here.

Reply:
Good point. It should be pre-order, and  iter_children should also be from first to last. Will update the text to reflect that.